From fa856e9d6cce8cac8f55bb00392725961c264b8f Mon Sep 17 00:00:00 2001
From: iliana destroyer of worlds <iweller@amazon.com>
Date: Tue, 6 Aug 2019 17:37:19 +0000
Subject: [PATCH] setup: Add root device argument to grub-setup

This patch originates from the OpenWRT tree:
https://github.com/openwrt/openwrt/blob/65c8f2890ca4f41f5b933b5bc1e43de86cc1bd54/package/boot/grub2/patches/100-grub_setup_root.patch
---
 include/grub/util/install.h |  4 ++--
 util/grub-install.c         |  4 ++--
 util/grub-setup.c           | 12 +++++++++++-
 util/setup.c                | 10 +++++++---
 4 files changed, 22 insertions(+), 8 deletions(-)

[bcressey: updated for grub 2.06]
Signed-off-by: Ben Cressey <bcressey@amazon.com>

diff --git a/include/grub/util/install.h b/include/grub/util/install.h
index 51f3b13..18cd28f 100644
--- a/include/grub/util/install.h
+++ b/include/grub/util/install.h
@@ -205,13 +205,13 @@ grub_install_get_image_target (const char *arg);
 void
 grub_util_bios_setup (const char *dir,
 		      const char *boot_file, const char *core_file,
-		      const char *dest, int force,
+		      const char *root, const char *dest, int force,
 		      int fs_probe, int allow_floppy,
 		      int add_rs_codes, int warn_short_mbr_gap);
 void
 grub_util_sparc_setup (const char *dir,
 		       const char *boot_file, const char *core_file,
-		       const char *dest, int force,
+		       const char *root, const char *dest, int force,
 		       int fs_probe, int allow_floppy,
 		       int add_rs_codes, int warn_short_mbr_gap);
 
diff --git a/util/grub-install.c b/util/grub-install.c
index 5babc7a..4233ac8 100644
--- a/util/grub-install.c
+++ b/util/grub-install.c
@@ -1770,7 +1770,7 @@ main (int argc, char *argv[])
 	if (install_bootsector)
 	  {
 	    grub_util_bios_setup (platdir, "boot.img", "core.img",
-				  install_drive, force,
+				  NULL, install_drive, force,
 				  fs_probe, allow_floppy, add_rs_codes,
 				  !grub_install_is_short_mbrgap_supported ());
 
@@ -1801,7 +1801,7 @@ main (int argc, char *argv[])
 	if (install_bootsector)
 	  {
 	    grub_util_sparc_setup (platdir, "boot.img", "core.img",
-				   install_drive, force,
+				   NULL, install_drive, force,
 				   fs_probe, allow_floppy,
 				   0 /* unused */, 0 /* unused */ );
 
diff --git a/util/grub-setup.c b/util/grub-setup.c
index 1783224..48cde49 100644
--- a/util/grub-setup.c
+++ b/util/grub-setup.c
@@ -87,6 +87,8 @@ static struct argp_option options[] = {
    N_("install even if problems are detected"), 0},
   {"skip-fs-probe",'s',0,      0,
    N_("do not probe for filesystems in DEVICE"), 0},
+  {"root-device", 'r', N_("DEVICE"), 0,
+   N_("use DEVICE as the root device"), 0},
   {"verbose",     'v', 0,      0, N_("print verbose messages."), 0},
   {"allow-floppy", 'a', 0,      0,
    /* TRANSLATORS: The potential breakage isn't limited to floppies but it's
@@ -130,6 +132,7 @@ struct arguments
   char *core_file;
   char *dir;
   char *dev_map;
+  char *root_dev;
   int  force;
   int  fs_probe;
   int allow_floppy;
@@ -178,6 +181,13 @@ argp_parser (int key, char *arg, struct argp_state *state)
         arguments->dev_map = xstrdup (arg);
         break;
 
+      case 'r':
+        if (arguments->root_dev)
+          free (arguments->root_dev);
+
+        arguments->root_dev = xstrdup (arg);
+        break;
+
       case 'f':
         arguments->force = 1;
         break;
@@ -313,7 +323,7 @@ main (int argc, char *argv[])
   GRUB_SETUP_FUNC (arguments.dir ? : DEFAULT_DIRECTORY,
 		   arguments.boot_file ? : DEFAULT_BOOT_FILE,
 		   arguments.core_file ? : DEFAULT_CORE_FILE,
-		   dest_dev, arguments.force,
+		   arguments.root_dev, dest_dev, arguments.force,
 		   arguments.fs_probe, arguments.allow_floppy,
 		   arguments.add_rs_codes, 0);
 
diff --git a/util/setup.c b/util/setup.c
index 8b22bb8..960aeda 100644
--- a/util/setup.c
+++ b/util/setup.c
@@ -252,14 +252,13 @@ identify_partmap (grub_disk_t disk __attribute__ ((unused)),
 void
 SETUP (const char *dir,
        const char *boot_file, const char *core_file,
-       const char *dest, int force,
+       const char *root, const char *dest, int force,
        int fs_probe, int allow_floppy,
        int add_rs_codes __attribute__ ((unused)), /* unused on sparc64 */
        int warn_small)
 {
   char *core_path;
   char *boot_img, *core_img, *boot_path;
-  char *root = 0;
   size_t boot_size, core_size;
   grub_uint16_t core_sectors;
   grub_device_t root_dev = 0, dest_dev, core_dev;
@@ -311,7 +310,10 @@ SETUP (const char *dir,
 
   core_dev = dest_dev;
 
-  {
+  if (root)
+    root_dev = grub_device_open(root);
+
+  if (!root_dev) {
     char **root_devices = grub_guess_root_devices (dir);
     char **cur;
     int found = 0;
@@ -324,6 +326,8 @@ SETUP (const char *dir,
 	char *drive;
 	grub_device_t try_dev;
 
+	if (root_dev)
+	  break;
 	drive = grub_util_get_grub_dev (*cur);
 	if (!drive)
 	  continue;
-- 
2.21.3

